escrow/atomic_storage.go

```diff
@@ -69,6 +94,59 @@ func (storage *PrefixedAtomicStorage) Delete(key string) (err error) {
 	return storage.delegate.Delete(storage.keyPrefix + "/" + key)
 }
 
+func (storage *PrefixedAtomicStorage) StartTransaction(conditionKeys []string) (transaction Transaction, err error) {
+	return storage.delegate.StartTransaction(conditionKeys)
+}
+
+func (storage *PrefixedAtomicStorage) appendKeyPrefix(conditionKeys []string) (preFixedConditionKeys []string) {
+	preFixedConditionKeys = make([]string, len(conditionKeys))
+	for i, key := range conditionKeys {
+		preFixedConditionKeys[i] = storage.keyPrefix + "/" + key
+	}
+	return preFixedConditionKeys
+}
+
+func (storage *PrefixedAtomicStorage) appendKeyPrefixOn(update []*KeyValueData) []*KeyValueData {
+	for _, keyValue := range update {
+		keyValue.Key = storage.keyPrefix + "/" + keyValue.Key
```

Comment:
you are right , they keys are updated !, Would this become a problem say we have thousands of requests hitting Daemon when we copy to an array ?the update values are always overridden by newer values when the last attempt to persist fails and we retry

Reply:
> Would this become a problem say we have thousands of requests hitting Daemon when we copy to an array ?

Whether copying it is a perfomance hit? I think yes. Whether it is the most serious problem in our case? I am not sure, only performance testing can show. My understanding common and recommended practice in `Golang` is passing parameters by value instead of passing by reference, and `Golang` should make it effective enough. This is why I would suggest using `[]KeyValueData` instead of `[]*KeyValueData`. I almost sure that `[]KeyValueData` will work faster than `[]*KeyValueData`, because in case of `[]KeyValueData` you don't need explicitly allocate memory for each item.

The problem which bothers me in current implementation is that client cannot be sure that data which is passed to the `CompleteTransaction` call will stay unchanged after call is finished. Change in the data is unexpected by the client and I think it is not correct allowing it.

Reply:
makes sense @vsbogd , will address this

Reply:
lets use Pass by reference then every where consistently. 
say use the below signature for CompleteTransaction
CompleteTransaction(transaction Transaction, update []KeyValueData) (ok bool, err error)
why even Give an option for someone to modify the value by mistake

Reply:
Please see https://github.com/anandrgitnirman/snet-daemon/pull/46

Reply:
this is fixed now